### PR TITLE
Add support for LSTM and GRU Layers and add implemention all RNN layers using cuDNN

### PR DESCRIPTION
**Pull request by lmoneta**
This PR replaces #3924  by Susrya.  Integrates his code with the new developments (RTensor) for CPu and in addition extends his implmentation by : 
 - fixing backpropagation for GRU
 - add cuDNN implmentation for GPU for simple RNN, LSTM and GRU

### CODE REVIEW

File: tmva/tmva/test/DNN/GRU/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,52 @@
+############################################################################
+# CMakeLists.txt file for building TMVA/DNN/GRU tests.
+# @author Surya S Dwivedi
+############################################################################
+
+project(tmva-tests)
```
Comment (oshadura) [inline tmva/tmva/test/DNN/GRU/CMakeLists.txt]: @lmoneta, I wanted to ask why each directory in  tmva/tmva/test/DNN/[GRU, LSTM and others] is setting the same project "tmva-tests"? (it looks like copy-paste and is not needed at all). The same question is about next lines: 

`find_package(ROOT REQUIRED)
set(Libraries Core MathCore Matrix TMVA)
include_directories(${ROOT_INCLUDE_DIRS})` 
Are they really needed?

**Comment by lmoneta** [comment]
@phsft-bot  build!

File: math/mathcore/src/RichardsonDerivator.cxx [hunk 2]
```diff
@@ -15,6 +15,8 @@
 #include <algorithm>
 
 #include "Math/Error.h"
+#include <iostream>
```
Comment (stwunsch) [inline math/mathcore/src/RichardsonDerivator.cxx]: Remove `iostream` include if it's not needed

File: math/mathcore/src/RichardsonDerivator.cxx [hunk 3]
```diff
@@ -86,8 +88,9 @@ double RichardsonDerivator::Derivative1 (const IGenFunction & function, double x
    xx = x+h;     double f1 = (function)(xx);
 
    xx = x-h;     double f2 = (function)(xx);
-
+
```
Comment (stwunsch) [inline math/mathcore/src/RichardsonDerivator.cxx]: Fix whitespaces

File: tmva/tmva/inc/TMVA/DNN/Architectures/Cpu.h [hunk 4]
```diff
@@ -84,9 +85,11 @@ class TCpu
    using AlgorithmHelper_t       = DummyConvolutionBwdFilterAlgo;
    using AlgorithmDataType_t     = DummyDataType;
    using ReduceTensorDescriptor_t = DummyDataType;
+   using RecurrentDescriptor_t    = DummyDataType;
 
-   using EmptyDescriptor_t       = DummyDescriptor;        // Used if a descriptor is not needed in a class
+   using EmptyDescriptor_t = DummyDescriptor; // Used if a descriptor is not needed in a class
```
Comment (stwunsch) [inline tmva/tmva/inc/TMVA/DNN/Architectures/Cpu.h]: Breaks formatting of the header

File: tmva/tmva/inc/TMVA/DNN/Architectures/Cpu.h [hunk 5]
```diff
@@ -143,12 +165,20 @@ class TCpu
                                        TDescriptors * & /*descriptors*/,
                                        const DNN::CNN::TConvParams & /*params*/,
                                        PoolingLayer_t * /*L = nullptr*/) {}
+   static void  InitializeRNNWorkspace(TWorkspace *& /*workspace*/, TDescriptors *& /*descriptors*/, GenLayer_t * /*L*/) {}
```
Comment (stwunsch) [inline tmva/tmva/inc/TMVA/DNN/Architectures/Cpu.h]: double whitespace behind `static void`

**Pull request merged**